doc/manage-duniverse.mld

```diff
@@ -0,0 +1,116 @@
+{1 Considerations}
+
+When using opam-monorepo the question arises: how to manage it in comparison to
+the regular opam workflow? Compared to regular opam usage using opam-monorepo
+brings lockfiles and a duniverse folder into the repository. There are multiple
+ways to handle this, with multiple advantages and disadvantages. This document
+explains the considerations behind dealing with each of those, along with our
+recommendations.
+
+{1 Methods}
+
+We assume the opam file is committed to the repository. Running [opam monorepo
+lock] will create a [.locked] opam file in addition to the original opam file.
+Calling [opam monorepo pull] will create a [duniverse] folder into which it
+will unpack the sources of the packages that are mentioned in the lockfile.
+
+{2 Not locked}
+
+While this is not a [opam-monorepo] workflow per se, projects that do not use
+[opam-monorepo] can still be used with [opam-monorepo]. As such every
+contributor is responsible for creating and maintianing their own lockfile and
+duniverse.
+
+Advantages include:
+
+{ul
+{- No opt-in to [opam-monorepo] required, every developer can chose to use it
+   or not}
+}
+
+Unfortunately, in this approach the disadvantages dominate:
+
+{ul
+{- Everyone might end up with different packages in their duniverse, thus low
+   reproducability}
+{- Everyone needs to have opam-monorepo installed}
+}
+
+The [opam-monorepo] maintainers don't recommend this approach, as is has hardly
+any advantages over an [opam-monorepo]-less approach.
+
+{2:lockfile-in-git Lockfile in Git}
+
+Another approach is very similar to the regular opam workflow, but in addition
+to having the [opam] file committed to the repository, also the lockfile that
+[opam monorepo lock] generates is added.
+
+Advantages include:
+
+{ul
+{- The git repostitory does not grow by much, only one file}
+{- The packages to be installed are reproducible}
+{- Installation on the packages regularly into the Opam switch is still
+   possible}
+}
+
+Disadvantages might be
+
+{ul
+{- Everyone needs to have opam-monorepo installed to pull the dependencies}
+}
+
+[opam-monorepo] maintainers encourage people to adopt this pattern, since it
+provides a good practical compromise between reproducibility and
+practicability.
+
+{2:duniverse-in-git Duniverse in Git}
+
+The most extreme approach is to create a lockfile with [opam monorepo lock],
+download all the dependencies via [opam monorepo pull] and then commit both the
+lockfile as well as the pulled sources.
+
+This appraoch has some advantages:
+
+{ul
+{- Only the person that creates the lockfile and duniverse must have
+   [opam-monorepo] installed}
+{- The repository can be immediately used after cloning as everything is
+   self-contained}
+{- Can be used offline}
```

Comment:
another point here (which is both good and bad and you listed it only in the bad part :p): it's easy to quickly patch dependencies when you are trying to work collaboratively in a large-scale change. So it's a great way to experiment with new features like eio, or to test new releases, etc.

Reply:
Isn't this covered by:
```
{- Easy to modify dependencies if required}
```

It might have been added after your initial comment, I'm mostly trying to figure out whether this comment is addressed or not.

Reply:
I think I added this note later on after seeing @samoht's comment, just didn't mark as resolved because I was pondering pros and cons of doing so.

I think this workflow with "ejecting" changes is certainly interesting, but I am not sure using opam-monorepo (at least in its current form) is the optimal approach because it would require a complicated system track the changes. Especially given what opam-monorepo usually unpacks are tarballs, one would need to extract the changes, correlate them back to the git repo (which might not correspond to the tarball and also might have seen development since that particular release).

To me personally it seems more manageable to symlink or pin-depend or git-submodule a repo for iterating on changes to the dependencies than to modify files that opam-monorepo has vendored. That way it is much easier to differentiate between upstream and my changes and much easier to submit those upstream or have them in my own fork.